File: 239.sliding-window-maximum.cpp
/*
 * @lc app=leetcode id=239 lang=cpp
 *
 * [239] Sliding Window Maximum
 */

// @lc code=start
#include <vector>
using std::vector;
#include <deque>
using std::deque;

class MyQueue {
    public:
        void pop(int x){
            if (!dq.empty() && dq.front() == x)
                dq.pop_front();
        }

        void push(int x){
            while (!dq.empty() && x > dq.back())
                dq.pop_back();
            dq.push_back(x);
        }

        int front() {
            return dq.front();
        }
    private:
        deque<int> dq;
};

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        MyQueue q;
        for (int i = 0; i != nums.size() - k + 1; ++i) {
            if (!i) {
                for (int j = 0; j != k; ++j){
                    q.push(nums[i+j]);
                }
            } else {
                q.pop(nums[i-1]);
                q.push(nums[i+k-1]);
            }
            res.push_back(q.front());
        }
        return res;
    }
};
// @lc code=end

File: 239.sliding-window-maximum.py
#
# @lc app=leetcode id=239 lang=python3
#
# [239] Sliding Window Maximum
#
# https://leetcode.com/problems/sliding-window-maximum/description/
#
# algorithms
# Hard (40.66%)
# Likes:    2608
# Dislikes: 154
# Total Accepted:    224.5K
# Total Submissions: 550.3K
# Testcase Example:  '[1,3,-1,-3,5,3,6,7]\n3'
#
# Given an array nums, there is a sliding window of size k which is moving from
# the very left of the array to the very right. You can only see the k numbers
# in the window. Each time the sliding window moves right by one position.
# Return the max sliding window.
#
# Example:
#
#
# Input: nums = [1,3,-1,-3,5,3,6,7], and k = 3
# Output: [3,3,5,5,6,7]
# Explanation:
#
# Window position                Max
# ---------------               -----
# [1  3  -1] -3  5  3  6  7       3
# ⁠1 [3  -1  -3] 5  3  6  7       3
# ⁠1  3 [-1  -3  5] 3  6  7       5
# ⁠1  3  -1 [-3  5  3] 6  7       5
# ⁠1  3  -1  -3 [5  3  6] 7       6
# ⁠1  3  -1  -3  5 [3  6  7]      7
#
#
# Note:
# You may assume k is always valid, 1 ≤ k ≤ input array's size for non-empty
# array.
#
# Follow up:
# Could you solve it in linear time?
#
from typing import List
# @lc code=start
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        maxValues = []
        if not nums:
            return maxValues
        maxValue = max(nums[:k])
        maxValues.append(maxValue)
        prevLeft = 0
        right = k
        while right != len(nums):
            if nums[prevLeft] == maxValue:
                maxValue = max(nums[prevLeft+1: right+1])
            elif nums[right] > maxValue:
                maxValue = nums[right]
            maxValues.append(maxValue)
            prevLeft += 1
            right += 1
        return maxValues

# @lc code=end
if __name__ == "__main__":
    nums = [1,3,-1,-3,5,3,6,7]
    k = 3
    print(Solution().maxSlidingWindow(nums, k))

